Classes/CacheHelper.php

```diff
@@ -42,26 +43,39 @@ public function __construct(FrontendInterface $cache = null)
      * @param callable $loader
      * @return array
      */
-    public function get(string $cacheIdentifier, callable $loader): array
+    public function get(string $cacheIdentifier, callable $loader, ContentObjectRenderer $contentObjectRenderer): array
     {
         $pages = $this->cache->get($cacheIdentifier);
         if (is_array($pages)) {
+            $this->buildTagsAndAddThemToPageCache($pages, $contentObjectRenderer);
             return $pages;
         }
 
         // Do the actual work
         $pages = $loader();
 
         // Calculate tags + lifetime
-        $usedPageIds = $this->getAllPageIdsFromItems($pages);
-        $tags = array_map(function ($pageId) {
-            return 'pageId_' . $pageId;
-        }, $usedPageIds);
+        $tags = $this->buildTagsAndAddThemToPageCache($pages, $contentObjectRenderer);
         $maximumLifeTime = $this->getMaxLifetimeOfPages($pages, $GLOBALS['TSFE']->get_cache_timeout());
         $this->cache->set($cacheIdentifier, $pages, $tags, $maximumLifeTime);
         return $pages;
     }
 
+    /**
+     * @param mixed[] $pages
+     * @param ContentObjectRenderer $contentObjectRenderer
+     * @return string[]
+     */
+    protected function buildTagsAndAddThemToPageCache(array $pages, ContentObjectRenderer $contentObjectRenderer): array
+    {
+        $usedPageIds = $this->getAllPageIdsFromItems($pages);
+        $tags = array_map(function ($pageId) {
+            return 'menuId_' . $pageId;
+        }, $usedPageIds);
+        $contentObjectRenderer->stdWrap_addPageCacheTags('', ['addPageCacheTags' => implode(',', $tags)]);
```

Comment:
I would recommend handing over TSFE everywhere instead and use `$TSFE->addCacheTags()` instead, so you don't need to do implode/explode anymore